rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,583 @@
+excluded from compatibility compliance. Note that generated pipeline code &
+config are considered user code, therefore they can expect the same
+compatibility compliance as listed above.
+
+Putting it differently:
+
+*   The structure of code and configs generated is subject to (possibly
+    breaking) change between minor versions;
+*   Once generated, the produced pipeline code and configs are essentially user
+    code which relies on TFX’s public API, thus guaranteed to work with newer
+    minor versions within the same major version, as explained in backward
+    compatibility rules.
+    *   This rule ensures that even if a user upgraded the `tfx` package in
+        their environment, the scaffolded pipeline can still function.
+
+#### Examples, Notebooks and Tutorials {#examples-notebooks-and-tutorials}
+
+Examples, Notebooks (Colab) and Tutorials are **excluded** from the
+compatibility compliance. These are created mostly for demonstrating how to use
+TFX in accordance with best practices, therefore often subject to a particular
+version of TFX.
+
+These examples should use the stable public API of `tfx` as much as possible.
+Usage of experimental or deprecated APIs in these examples should be explicitly
+called out, and the examples should pin their dependency version of `tfx` down
+to a narrow range as much as possible in these rare cases.
+
+#### New Orchestrator {#new-orchestrator}
+
+The new orchestrator is in early stages of development and we don’t think it’ll
+be production-ready for 1.0 timeline, thus it will enter 1.0 as `experimental`.
+
+#### Portability Layer to Other Runners (aka Orchestrators) {#portability-layer-to-other-runner-orchestrators}
+
+This portability layer in tfx refers to various DagRunners which runs TFX
+pipelines on various external orchestrators, the _portable_ module and the
+abstraction between CLI and Dag runner. These libraries will comply with
+backward compatibility.
+
+Notice that several Runners are experimental by their nature, or slated to be
+deprecated. Specifically, we plan to support the following runners with backward
+compatibility:
+
+*   LocalDagRunner
+*   KubeflowV2DagRunner
+
+The following Dag runners will remain experimental:
+
+*   AirflowDagRunner
+    *   Honestly speaking, support of this one is on the fence. Due to lack of
+        proper configuration for a distributed backend support, it’s not much
+        more powerful than the LocalDagRunner when used out-of-box. However,
+        some partner teams managed to connect this to a distributed backend.
+        Until that is supported in TFX, we don’t think this one is stable.
+
+The following dag runners are slated to be deprecated and therefore won’t
+comply:
+
+*   BeamDagRunner: this will be replaced by the LocaDagRunner
+*   KubeflowDagRunner: this dag runner implementation based on KFP v1 API will
+    be replaced by the V2 dag runner above.
+
```

Comment:
Where does InteractiveContext fit here?

Reply:
https://www.tensorflow.org/tfx/api_docs/python/tfx/orchestration/experimental/interactive/interactive_context/InteractiveContext

Reply:
So we should not expect InteractiveContext to graduate from experimental on TFX 1.X?

Reply:
No we are not. We don't want to rush the graduation of experimental features, and this one hasn't picked up enough internal or production uses yet to qualify our graduation standard.

Reply:
I see. Would be interesting to know about graduation standard for transparency. We have quite large internal usage of this.